README.md

```diff
@@ -1,61 +1,419 @@
 # Abstract
 
-Basic definitions for abstract algebraic structures in Swift.
+A take on abstract algebraic structures, in Swift.
 
-To each structure are assigned some laws, that can be verified with property-based testing (and are verified for all the types considered in this repo using SwiftCheck), so than anyone can give some structure to more types: the only thing that's required is to verify the laws.
+------
+
+`Abstract` is a Swift library that defines protocols for the main [abstract algebraic structures](https://en.wikipedia.org/wiki/Abstract_algebra), along with some concrete implementations for them that declare specific semantics.
 
-Files have comments to help understand the underlying theory, but each file builds on top of the others, so a good order to read them could be:
+The library also provides tools to test the concrete types for the axioms required by each algebraic structure: tests can then be performed by property-based testing libraries like [SwiftCheck](https://github.com/typelift/SwiftCheck).
 
-- Law.swift
-- Magma.swift
-- Semigroup.swift
-- Monoid.Swift
-- CommutativeMonoid.swift
-- BoundedSemilattice.swift
+------
 
-The other files can be read in any order after these 6.
+## System Requirements
 
-In the `Utilities` folder there are some useful functions, to start using these types in real-world programming.
+`Abstract` supports macOS 10.10+ and iOS 8.3+.
 
 ------
 
-This work is largely based on some talks of the 2017 [Functional Swift Conference](http://2017.funswiftconf.com), in particular [this](https://www.youtube.com/watch?v=6z9QjDUKkCs) and [this](https://www.youtube.com/watch?v=VFPhPOnPiTY).
+## Setup
+
+`Abstract` is compatible with [Carthage](https://github.com/Carthage/Carthage): please refer to Carthage documentation for how to add `Abstract` as a dependency of your project.
+
+### SwiftPM
+Please add this line to your `Package.swift` file's dependencies section:
 
-The goal is to try and give a pragmatic, practical standard for abstract algebra usage in Swift. My main point is also that a usable library should expose a way to enforce the laws on types, and that's what I tried to do here with the `Law` namespace.
+```
+.package(url: "https://github.com/typelift/Abstract.git",
+         from: Version(0,0,0))
+```
 
-There's still a lot of limitations in Swift's' type system, that require some overrides and workarounds to be able to express some things (particularly when working with functions). I will update the library along with Swift's evolution, let's hope that those loose ends are tied as soon as possible: for example when [this](https://github.com/apple/swift-evolution/blob/master/proposals/0143-conditional-conformances.md) (accepted) proposal will be implementaed, everything will change for good, especially in the standard library.
+Also add `"Abstract"` to your target's dependencies, an also to your testTarget dependencies if you want to test your types with the `Law` functions.
```

Comment:
> Also add `"Abstract"` to your target's dependencies, an also to your testTarget dependencies if you want to test your types with the `Law` functions.

To use the structures in this library, add `"Abstract"` to your target's dependencies.  To additionally test algebraic laws with the framework, add `"Abstract"` as a dependency to the relevant `testTarget`s.